File: app/src/test/java/com/permissionnanny/content/ProxyContentProviderTest.java
package com.permissionnanny.content;

import android.content.pm.ProviderInfo;
import android.database.Cursor;
import android.net.Uri;
import com.permissionnanny.NannyAppTestRunner;
import com.permissionnanny.common.test.NannyTestCase;
import com.permissionnanny.lib.request.RequestParams;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.robolectric.RuntimeEnvironment;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;

@RunWith(NannyAppTestRunner.class)
public class ProxyContentProviderTest extends NannyTestCase {

    ProxyContentProvider mProxyContentProvider;
    RequestParams mRequestParams;
    Uri mUri;
    @Mock ProviderInfo mProviderInfo;

    @Before
    public void setUp() throws Exception {
        mRequestParams = new RequestParams();
        mUri = Uri.parse("content://authority/123");
        mProxyContentProvider = new ProxyContentProvider();
        mProxyContentProvider.attachInfo(RuntimeEnvironment.application, mProviderInfo);
    }

    @Test
    public void query() throws Exception {
        ProxyContentProvider.approvedRequests.put(123L, mRequestParams);

        Cursor ans = mProxyContentProvider.query(mUri, null, null, null, null);

        assertThat(ans, notNullValue());
    }

    @Test
    public void query_shouldReturnNull_whenDuplicateRequests() throws Exception {
        ProxyContentProvider.approvedRequests.put(123L, mRequestParams);

        Cursor ans = mProxyContentProvider.query(mUri, null, null, null, null);
        Cursor ans1 = mProxyContentProvider.query(mUri, null, null, null, null);

        assertThat(ans, notNullValue());
        assertThat(ans1, nullValue());
    }

    @Test
    public void query_shouldReturnNull_whenUnapprovedRequest() throws Exception {
        Cursor ans = mProxyContentProvider.query(mUri, null, null, null, null);

        assertThat(ans, nullValue());
    }

    @Test
    public void query_shouldReturnNull_whenInvalidRequest() throws Exception {
        Cursor ans = mProxyContentProvider.query(Uri.parse("content://authority/abc"), null, null, null, null);

        assertThat(ans, nullValue());
    }

    @Test
    public void insert() throws Exception {
        Uri ans = mProxyContentProvider.insert(null, null);

        assertThat(ans, nullValue());
    }

    @Test
    public void delete() throws Exception {
        int ans = mProxyContentProvider.delete(null, null, null);

        assertThat(ans, is(0));
    }

    @Test
    public void update() throws Exception {
        int ans = mProxyContentProvider.update(null, null, null, null);

        assertThat(ans, is(0));
    }
}

File: app/src/test/kotlin/com/permissionnanny/content/ProxyContentProviderTest.kt
package com.permissionnanny.content

import android.content.pm.ProviderInfo
import android.net.Uri
import com.permissionnanny.NannyAppTestCase
import com.permissionnanny.lib.request.RequestParams
import org.hamcrest.Matchers.*
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.robolectric.RuntimeEnvironment

class ProxyContentProviderTest : NannyAppTestCase() {

    private lateinit var proxyContentProvider: ProxyContentProvider
    private lateinit var requestParams: RequestParams
    private lateinit var uri: Uri
    @Mock private lateinit var providerInfo: ProviderInfo

    @Before
    fun setUp() {
        requestParams = RequestParams()
        uri = Uri.parse("content://authority/123")
        proxyContentProvider = ProxyContentProvider()
        proxyContentProvider.attachInfo(RuntimeEnvironment.application, providerInfo)
    }

    @Test
    fun query() {
        ProxyContentProvider.approvedRequests.put(123L, requestParams)

        val ans = proxyContentProvider.query(uri, null, null, null, null)

        assertThat(ans, notNullValue())
    }

    @Test
    fun query_shouldReturnNull_whenDuplicateRequests() {
        ProxyContentProvider.approvedRequests.put(123L, requestParams)

        val ans = proxyContentProvider.query(uri, null, null, null, null)
        val ans1 = proxyContentProvider.query(uri, null, null, null, null)

        assertThat(ans, notNullValue())
        assertThat(ans1, nullValue())
    }

    @Test
    fun query_shouldReturnNull_whenUnapprovedRequest() {
        val ans = proxyContentProvider.query(uri, null, null, null, null)

        assertThat(ans, nullValue())
    }

    @Test
    fun query_shouldReturnNull_whenInvalidRequest() {
        val ans = proxyContentProvider.query(Uri.parse("content://authority/abc"), null, null, null, null)

        assertThat(ans, nullValue())
    }

    @Test
    fun insert() {
        val ans = proxyContentProvider.insert(null, null)

        assertThat(ans, nullValue())
    }

    @Test
    fun delete() {
        val ans = proxyContentProvider.delete(null, null, null)

        assertThat(ans, equalTo(0))
    }

    @Test
    fun update() {
        val ans = proxyContentProvider.update(null, null, null, null)

        assertThat(ans, equalTo(0))
    }
}
